pkg/trace/api/api.go

```diff
@@ -64,6 +64,37 @@ const (
 	// Traces: msgpack/JSON (Content-Type) slice of traces + returns service sampling ratios
 	// Services: deprecated
 	v04 Version = "v0.4"
+	// v05
+	// Traces: msgpack (Content-Type) dictionary + slice of traces + returns service sampling ratios
```

Comment:
@richardstartin can you please review this spec I just wrote? How does it look? Is this along the lines of what you were thinking? Is there anything I need to add or remove?

Reply:
Looks great except for conflating null with the empty string

Reply:
I corrected that to say that any top-level field can be `nil`. Other than the few top-level fields which are strings (Service, Name, Resource and Type) that can also be `nil`, no other string can be `nil` inside the `Meta` and `Metrics` map. I hope this helps.

I also added more detail in https://github.com/DataDog/datadog-agent/pull/5965#discussion_r459684609

Reply:
I think my last message might've been confusing (it is a little bit to me when I re-read it). I'll try to clarify. These are the span fields that we have today along with their types:

* Service   (string)
* Name      (string)
* Resource  (string)
* TraceID   (uint64)
* SpanID    (uint64)
* ParentID  (uint64)
* Start     (int64)
* Duration  (int64)
* Error     (int32)
* Meta      (map[string]string)
* Metrics   (map[string]float64)
* Type      (string)

This is the behaviour expected by the decoder:
* Any of the above fields can be set to `nil` when they are absent, regardless of their type, be it `string`, `int64`, `uint64`, `map[string]string` etc.
* When the `Meta` and `Metrics` maps are present, neither their keys nor values can ever be `nil`, only `string` (or `float64`).

What I'm proposing is that `nil` has no place in the dictionary because it is a valid value anywhere above, regardless of type. The dictionary is meant to be string only. I am *NOT* suggesting conflating empty string with `nil`. Sorry if I was unclear about that. I want `nil` fields to continue to be `nil` fields, without making exceptions for `string`, which would just confuse things in type definitions, code, and spec clarity.

In other words, if we have fields with the types `uint64`, `int64`, `int32`, `map[string]string` and `map[string]float64` which can be represented as `nil` on the wire, we want the `string` type to be represented in the same way, not as `int(0)` because:

1. There is no reason for special treatment.
2. It would make a clear string dictionary confusing by adding an exception.

Reply:
I don't think this issue is resolved. I quote what I said in the conversation which was marked as resolved:

The way I had imagined this working - and what we implemented and what I think we wanted to evaluate on the Java tracer side - was that all of the string values in the spans would be dictionary encoded, so always integers. To be transparent, I didn't know that the agent currently supported binary values, which is cool but not a feature we'd taken advantage of in the Java tracer.

I don't think this means we can't evaluate what we wanted to: if we allow the dictionary to be heterogeneously typed, potentially containing any type convertible to a string (msgpack nil, string, bin), we can constrain the keys and values in the messages to always be dictionary codes. This would mean that all of the runtime type/null checking could be done up front as the dictionary is unpacked, and make the bulk of the work - unpacking spans - as simple as possible.

And your immediate response:

This sounds good to me. Let's make sure we don't need bin (I'm pretty sure we don't) and go for replacing all strings everywhere with int which will be the index of the string in the dictionary. I was thinking the easiest, nicest implementation with less complex code would be to have only strings in the dictionary. Since nil means always empty string, it could be represented as "" in the dictionary.

At this point I thought you got it - the values in the message will be dictionary encoded, and there will be no nil values in the message at points you would need to dictionary decode, and I thought we just had to discuss how to represent missing values in the encoding itself; you would prefer the empty string and I prefer nil. Had I misunderstood where this conversation had got to?

Reply:
You're right. Sorry about that. There was a bit of confusion on my side when this discussion happened in the sense that I thought it would be possible to have `nil` values in place of `string` anywhere in the encoding, but it turns out this is only true for a few top-level string keys. I summarised the whole idea in my previous comment (https://github.com/DataDog/datadog-agent/pull/5965#discussion_r459859215). I'd like all top-level fields to treat `nil` equally and not have `string` get different treatment, unless there are good reasons to. A top-level string field that is absent can be replaced with 1 byte (`nil`). No need to add an `int(0)` there which points to a map of strings that has an exception at index 0 which is `nil`.

Reply:
OK. I understand your position. We disagree, but I don't think it helps for me to reiterate my position. I will leave this alone until there is quorum consensus in a wider group.

> A top-level string field that is absent can be replaced with 1 byte (nil). No need to add an int(0) there which points to a map of strings that has an exception at index 0 which is nil.

Just for the record, notwithstanding the above, it's not for the sake of space but for the (achievable) efficiency of the decoding.

Reply:
If we can prove this efficiency by means of a benchmark, posting the results using the benchcmp tool as I did above, I am more than happy to do it the way you suggest. I was not aware you are confident of such improvements. Happy to help guide you how to write a benchmark in Go.

Reply:
I implemented what I was describing [here](https://github.com/richardstartin/datadog-agent/commit/cf93514c921b42bb0a208b762c9cd25cde90937c), which actually continues to support binary strings too. The idea is that the null checks/type checks can be done up front when unpacking the dictionary. There's a positive difference, but the benchmark isn't really going to demonstrate the benefits of removing optionality/branching from a CPU bound workload because it doesn't generate enough data random enough to "surprise" the branch predictor. In practice, this probably won't matter because null values are very rare, and the branch predictor will be able to do a good job. 

I got these numbers on my system, which has a processor frequency of ~2.6GHz.

```
gbbr/v0.5 (this branch)

goos: linux
goarch: amd64
pkg: github.com/DataDog/datadog-agent/pkg/trace/pb
BenchmarkDecodeMsgArray
BenchmarkDecodeMsgArray-8   	  506625	      2352 ns/op	 122.04 MB/s	     528 B/op	      14 allocs/op


rgs/v0.5

goos: linux
goarch: amd64
pkg: github.com/DataDog/datadog-agent/pkg/trace/pb
BenchmarkDecodeMsgArray
BenchmarkDecodeMsgArray-8   	  551686	      2154 ns/op	 133.68 MB/s	     544 B/op	      14 allocs/op

```

I understand _some_ of the reservations about `nil` now; it complicates the creation of test cases, but this shouldn't be insurmountable if the testing were done against msgpack encoded payloads, where the msgpack type `nil` can be represented easily. The implementation on my branch doesn't assume that `nil` is at index zero, it just accepts it as a value in the dictionary, and replaces it with a default value.

Reply:
Note that at ~2us/span, we should be able to decode about 500,000 spans per second. Obviously not all of those spans will be sat in a few lines of L1 cache in real life, as the one span we have in this benchmark does, and there will be some unpredictability which leads to pipeline stalls, so it's going to be quite a lot slower, but **certainly** not over 10x slower. So I doubt we're focussing on the bottleneck here given that we're looking at baselining against a figure like 32k spans/second.

Reply:
Thanks for doing that. On thing is that in your branch, you have done more than just adding the nil exception for strings. You have also removed the for loop and the nil check for every other top level field (which isn’t string). That is what brought you the slight speed increase (I have already posted this benchmark in this PR).

Reply:
This is just the decoding logic I had imagined when I defined the encoding. My aim was to demonstrate how easy it would be for you to accommodate null and binary values in the dictionary if you wanted to, and that it would not slow the decoding down to offer this flexibility. It allows you to deduplicate the null and type checking as well as the strings themselves. So it might help to run a benchmark where there isn't just one span being decoded over and over again.

On an unrelated note, I would have been tempted not to make the numeric fields nullable; they all have very well defined default values, and the target format is protobuf, which excludes any field with a default value from the message anyway.

Reply:
Regarding the second part of your comment, it has crossed my mind too: disallowing nil everywhere. We have already proven that this does offer some performance gain, at least as far as the trace-agent goes. If we were to do that, there would be no need for nil anywhere, including the dictionary. It wouldn’t be a valid value anywhere. Today a string can only be nil on the 3 top-level fields. 

The only thing that is slightly stopping me from pursuing this idea is the fact that we know well that allocating an empty map on the client side has a cost which can be problematic. It can of course be worked around by adding conditions in the encoding code, but that means custom encoding code and I am not sure how easy it is to do on the client side in most languages. 

Nevertheless, I wanted to propose something that might simplify both of our work. We should maybe reduce the scope of this PR to simply implementing what is already there in v0.5, but with an added dictionary. The reason I am proposing this is because I am hoping it will unblock it.

Reply:
You don't need to allocate on the client side to write an empty map into a messagepack stream. You just write the byte `0x80` which is `FIXMAP | 0`. See the [spec](https://github.com/msgpack/msgpack/blob/master/spec.md#map-format-family). 

Is there something objectively problematic about the dictionary unpacking code from my branch below? 

```go
          for i := range dict {
               nextType, err := dc.NextType()
		if err != nil {
			return err			
		}			
		dict[i] = str	
		switch nextType {
		case msgp.NilType:		
			dict[i] = ""
			break
		case msgp.BinType:
			bytes, err := dc.ReadBytes(nil)
			if err != nil {
				return err
			}
			dict[i] = msgp.UnsafeString(bytes)
			break
		case msgp.StrType:
			utf8, err := dc.ReadString()
			if err != nil {
				return err
			}
			dict[i] = utf8
			break
		default:
			return fmt.Errorf("dictionary value at index %d has unsupported type", i)
		}
```

Reply:
The issue is that there is no need to de-duplicate `nil`. You have mentioned efficiency but your benchmark encompasses more changes than exclusively the one you were referring to. Currently only `Service`, `Name`, `Resource` and `Type` can be `nil` as strings.

The scope of this PR is:
* Remove field names from encoding (done)
* Introduce a dictionary to de-duplicate strings in meta and metrics.

The constraint is keeping all the functionality we had the same. If any of the current functionality is up for discussion or change, that's fine, but IMHO it should be a separate discussion in another PR.

Reply:
> The issue is that there is no need to de-duplicate nil. You have mentioned efficiency but your benchmark encompasses more changes than exclusively the one you were referring to. Currently only Service, Name, Resource and Type can be nil as strings.

I partially accept this point, but think we're a little eager to explain all of the difference in terms of simplifying the message unpacking and removing `nil` checks, and it's also been ignored that my patch maintains support for binary strings, which had been thrown out for efficiency on this branch. I added `nil` checking back in for all numeric fields [here](https://github.com/DataDog/datadog-agent/commit/17a12d743a50308cd1e8c3aca559e7506cf0dd3f) and it didn't seem to cost anything in this benchmark. Adding null checking for the maps `meta` and `metrics` actually regressed around 4%. The benchmark isn't really representative of the costs involved though. It would be more revealing to benchmark a typical payload consisting of ~40k traces with, say, 1000 null strings, and compare the difference between doing 1 null check during unpacking the dictionary vs doing 1000 of them unpredictably.

>The scope of this PR is:
>
>Remove field names from encoding (done)
>Introduce a dictionary to de-duplicate strings in meta and metrics.
>
> The constraint is keeping all the functionality we had the same. If any of the current functionality is up for discussion or change, that's fine, but IMHO it should be a separate discussion in another PR.

But you aren't doing this here - you're also dropping support for binary strings! You could maintain support for them if you just unpacked the dictionary the way I'm asking you to.

Reply:
At [e267aafcdebd6adff87928a3324e812f7da1b7d9](https://github.com/DataDog/datadog-agent/commit/e267aafcdebd6adff87928a3324e812f7da1b7d9) I put simpler `nil` checks in than I had tried and got this (with support for missing strings and binary strings):

```
goos: linux
goarch: amd64
pkg: github.com/DataDog/datadog-agent/pkg/trace/pb
BenchmarkDecodeMsgArray
BenchmarkDecodeMsgArray-8   	  471408	      2172 ns/op	 132.60 MB/s	     544 B/op	      14 allocs/op
PASS
```

Compared to `570f276a375b202c8a2319de43487357315d3013` (HEAD on this branch)
```
goos: linux
goarch: amd64
pkg: github.com/DataDog/datadog-agent/pkg/trace/pb
BenchmarkDecodeMsgArray
BenchmarkDecodeMsgArray-8   	  516584	      2352 ns/op	 122.03 MB/s	     528 B/op	      14 allocs/op
PASS
```

Can we still dismiss this as an unfair test? All numbers are `nil` checked, all strings are `nil` checked (via the single `nil` check done unpacking the dictionary!), all maps are `nil` checked. The only thing left is that the loop was eliminated, but it has to be to do this.

Reply:
> I partially accept this point, but think we're a little eager to explain all of the difference (~10% improvement) in terms of simplifying the message unpacking and removing nil checks [...]

All I was saying was that we were discussing that `nil` in the dictionary would be more efficient and to prove that it would be best to have a benchmark with only that change. I did not ignore any of your other work.

> Adding null checking for the maps meta and metrics actually regressed around 4% (on a 10% improvement, so still 6% better) [...]

That's interesting. I generally value simplicity and readability most, and performance second (most of the time). I guess there needs to be a balance. In this case, we are pursuing performance. What I'm worried about with removing `nil` for maps is that it requires more custom encoding code on both ends: the client will have to conditionally check that if a map is `nil`, `nil` is not written onto the stream and instead a 1-byte map header is, and the `trace-agent` will have to do the same: if an empty map is received, do not allocate one. This is hacky in multiple places and that's why I'd prefer to keep the `nil`.

I would like to re-emphasise the main goal I mentioned above in https://github.com/DataDog/datadog-agent/pull/5965#discussion_r460751336. We are getting into a lot of lateral discussions which will prolong this more and more. I'd suggest we make the core of the changes which we've already proven to provide benefit and keep everything else like it was before. If we don't, discussions will get more and more complicated with more and more considerations. After we make the core of the changes, I am happy to take time and discuss these other changes too which can be added on later. These parts are important and tricky because they involve other languages too and it would be nice to have other people talk through them.

> [...] and it's also been ignored that my patch maintains support for binary strings, which had been thrown out for efficiency on this branch.

Not at all! The current PR supports `bin` entries in the dictionary. The code which decodes the dictionary calls `parseString` which is unchanged and will support both `bin` and `string`, as we always did! Removing `bin` was just an experiment in a previous comment to see if it would make a difference. It is fully supported in the PR!

Reply:
> Can we still dismiss this as an unfair test? All numbers are nil checked, all strings are nil checked (via the single nil check done unpacking the dictionary!), all maps are nil checked. The only thing left is that the loop was eliminated, but it has to be to do this.

The thing is that now any number or string can be nil. Previously only the top level fields could. If the loop is the problem, we can take the current code and write it inline, but we've already tried that and it made no difference. Also consider that you forgot to `dc.ReadNil` when the value actually _is_ `nil`.

I feel we're getting lost here. The current PR supports binary strings and everything that we did before, implementing the spec in the "Description". Can you please outline again what changes you would prefer we make to this PR and why? I'm a bit lost.

Reply:
@gbbr it's [10% faster](https://github.com/DataDog/datadog-agent/pull/5965#discussion_r460806683) even with `nil` checks on all the non string types. Can you please just put a check for `NilType` into `parseString` so we can just move  forwards?

Reply:
I invite you to put a check for `NilType` into `parseString` based on this exact PR without any other changes and show that it's faster and I'm happy to do it. This is all I'm asking for since the beginning. Your current PR has [a lot more changes](https://github.com/DataDog/datadog-agent/compare/gbbr/v0.5...e267aafcdebd6adff87928a3324e812f7da1b7d9) than just that. Even if we want to do those changes, I'd like us to focus on the problem at hand exclusively.

Reply:
In your branch you have also added support for any value anywhere to be `nil`, not just on the top-level. We can't and shouldn't have that. What is a map key gets sent as `nil`? The trace-agent will crash. Why do we want extra branches there now? I'm re-iterating again to keep the current behaviour and focus on the main topics:
* Remove string fields
* De-duplicate strings

And leave everything else for other PRs.

Reply:
Since we've both proven that decoding and checking for `nil` _does cost_, I'd be fine with removing the possibility of `nil` on the top-level fields. There would be no incompatibility with any part of the product if we did that. Then we would gain the speed we're after. This means that no value can ever be `nil`, whether string or otherwise. I agree that could be a good change, despite the "hacky" map exceptions.

Reply:
An alternative idea is to open a Draft PR on top of my branch so I can see exactly the changes you want and we can discuss there. It might help me understand your point better and clarify any missing points? We’ll also get more people to chip in

Reply:
I am resolving this conversation to reset the discussion. The latest commit represents the implementation we've discussed internally, but can continue to be subject to discussion. I am hoping you will find (like me) that this has the best of both our views.